features/ForceExpiration.php

```diff
@@ -54,41 +55,46 @@ public function __construct(BetterPasswordPlugin $plugin) {
 	 * Register callback to detect if the user password has expired
 	 */
 	private function _addPasswordExpirationCheck() : void {
-		foreach (['LoadHandler', 'LoadComponentHandler', 'TemplateManager::display'] as $hook) {
-			Hook::add('LoadHandler', function ($hook, $args) {
-				if ($this->_isProcessed) {
-					return;
-				}
-				$this->_isProcessed = true;
-				$user = Application::get()->getRequest()->getUser();
-				$username = $_POST['username'] ?? null;
-				if (!$user) {
-					return;
-				}
-
-				if (!$this->_isPasswordExpired($user)) {
-					$this->_handleNotification($user);
-					return;
-				}
-
-				if (!$user->getMustChangePassword()) {
-					$user->setMustChangePassword(true);
-					Repo::user()->edit($user);
-				}
-			});
+		$user = Application::get()->getRequest()->getUser();
+		if (!$user) {
+			return;
+		}
+
+		$sessionManager = SessionManager::getManager();
+		$session = $sessionManager->getUserSession();
+		$user = $session->getUser();
+		if ($user) {
+			if (!$this->_isPasswordExpired($user)) {
+				$session->setSessionVar('betterPassword::lastNotification', true);
+				$this->_handleNotification($user, $session);
+				return;
+			}
+			else {
+				$user->setMustChangePassword(true);
+				Repo::user()->edit($user);
+			}
+		}
+		else {
+			return;
 		}
 	}
 
 	/**
 	 * Retrieve the password expiration date
 	 * @param User $user
-	 * @return DateTime Most recent time of password change
+	 * @return DateTime Expiration date
 	 */
 	private function _getExpirationDate(\PKP\user\User $user) : \DateTime {
 		/* @var $storedPasswordsDao StoredPasswordsDAO */
 		$storedPasswordsDao = DAORegistry::getDAO('StoredPasswordsDAO');
-		$storedPasswords = $storedPasswordsDao->newDataObjects($user->getId());
-		$mostRecent = new \DateTime ($storedPasswordsDao->getMostRecent($storedPasswords));
+		$storedPasswords = $storedPasswordsDao->getByUserId($user->getId());
+		if (empty($storedPasswords)) {
```

Comment:
This is a dramatic side effect: setting the last changed time to `now()` if no entry exists on a `get`.  A `get` should almost never `set` data, and if it does it should be documented and rationalized.

The election of using `now()` as the last change time is questionable.  Why `now()`?